File: linked_list/c/ex_3.c
/**
 * Delete Middle Node: Implement an algorithm to delete a node in the middle
 * (i.e., any node but the first and last node, not necessarily the exact
 * middle) of a singly linked list, given only access to that node.
 */

#include <stdio.h>
#include "linked_list.h"

void delete_middle_node(node *nd);

int main() {
    list *test_list = create_list();
    insert_multiple(test_list, 5, 1, 2, 3, 4, 5);
    node *middle = find_node(test_list, 3);
    delete_middle_node(middle);
    print_list(test_list);
}

void delete_middle_node(node *nd) {
    if (nd == NULL) {
        return;
    }

    node *next = nd->next;
    nd->data = next->data;
    nd->next = next->next;
    free(next);
} 

File: go/linked_lists/ctci/ex3.go
package ctci

import "errors"

// 1.3 Delete the Middle Node: Implement an algorithm to delete a node in the
// middle (i.e. any node but the first and last node) of a singly linked list,
// given access only to that node.

// DeleteMiddle "deletes" a middle node by copying the data of its next node
// into itself and deleting the next node.
//
// Time complexity: O(1)
// Space complexity: O(1)
func DeleteMiddle(s *SNode) error {
	if s == nil || s.Next == nil {
		return errors.New("s cannot be tail node or nil")
	}

	next := s.Next
	s.Data = next.Data
	s.Next = next.Next
	return nil
}
